pkg/server/singleprocess/auth.go

```diff
@@ -533,6 +533,20 @@ func (s *Service) GenerateRunnerToken(
 		}
 	}
 
+	decodedToken := s.decodedTokenFromContext(ctx)
+
+	switch k := decodedToken.Kind.(type) {
+	case *pb.Token_Runner_:
+		// Prevent a runner token from generating a token for a runner with an
+		// ID different from the ID encoded in the runner token
+		if k.Runner.Id != req.Id {
```

Comment:
Looks great to me! We should also check to make sure that `k.Runner.Id` would never be `nil`, and also add a test for this new method.